llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -1809,6 +1809,15 @@ bool RISCVTargetLowering::getTgtMemIntrinsic(IntrinsicInfo &Info,
   case Intrinsic::riscv_masked_atomicrmw_umax_i32:
   case Intrinsic::riscv_masked_atomicrmw_umin_i32:
   case Intrinsic::riscv_masked_cmpxchg_i32:
+  case Intrinsic::riscv_masked_atomicrmw_xchg_i64:
+  case Intrinsic::riscv_masked_atomicrmw_add_i64:
+  case Intrinsic::riscv_masked_atomicrmw_sub_i64:
+  case Intrinsic::riscv_masked_atomicrmw_nand_i64:
+  case Intrinsic::riscv_masked_atomicrmw_max_i64:
+  case Intrinsic::riscv_masked_atomicrmw_min_i64:
+  case Intrinsic::riscv_masked_atomicrmw_umax_i64:
+  case Intrinsic::riscv_masked_atomicrmw_umin_i64:
+  case Intrinsic::riscv_masked_cmpxchg_i64:
     Info.opc = ISD::INTRINSIC_W_CHAIN;
     Info.memVT = MVT::i32;
```

Comment:
I don't entirely understand why these intrinsics aren't different to the i32 versions - maybe I'm missing something here.

What I understand:
- smaller-than-32-bit accesses are are transformed into the `_i64` versions on riscv64.
- these are then expanded into sequences containing `lr.w.<order>`, and a bunch of masking

So what does the `_i64` in the name relate to? What has type `i64`?

I think I missed all the masked atomics work at some point, probably when I was working on another architecture, so sorry if these questions are dumb.

Reply:
I don't know if you've seen it yet, but I also posted https://github.com/llvm/llvm-project/pull/154845 to merge these using type overloading.